src/Servers/HttpSys/src/RequestProcessing/ResponseBody.cs

```diff
@@ -127,8 +127,11 @@ private unsafe void FlushInternal(bool endOfRequest, ArraySegment<byte> data = n
             if (endOfRequest && _leftToWrite > 0)
             {
                 _requestContext.Abort();
-                // This is logged rather than thrown because it is too late for an exception to be visible in user code.
-                LogHelper.LogError(Logger, "ResponseStream::Dispose", "Fewer bytes were written than were specified in the Content-Length.");
+                if (!RequestContext.DisconnectToken.IsCancellationRequested)
```

Comment:
This check will always give the same result because Abort cancels the token, you'd never get the log now.
https://github.com/aspnet/AspNetCore/blob/78b9b2e23e7709cef6659b11e7e0b266f7a1089f/src/Servers/HttpSys/src/RequestProcessing/RequestContext.cs#L172
At the very least it should be reversed to log before Aborting.
You also need to consider the lazy creation logic for DisconnectToken. If nobody requested the token before it was cancelled you might end up with an uncancellable token.
https://github.com/aspnet/AspNetCore/blob/78b9b2e23e7709cef6659b11e7e0b266f7a1089f/src/Servers/HttpSys/src/RequestProcessing/RequestContext.cs#L47
There might be room for improvement there, such as ensuring a token created after cancellation is created in the canceled state.

Reply:
I put the logging before Abort and as you said I need to make some changes on DisconnectToken.

I used this code

```
CancellationTokenSource cts = new CancellationTokenSource();
string address;
using (Utilities.CreateHttpServer(out address, async httpContext =>
{
    httpContext.Response.Headers["Content-length"] = " 30 ";
    await httpContext.Response.Body.WriteAsync(new byte[10], 0, 10);
    cts.Cancel();
}))
{
    using (HttpClient client = new HttpClient())
    {
        await client.GetAsync(address, cts.Token);
    }
}
```
to hit the FlushInternal and check DisconnectToken. In this scenario if I skip this if statement 
https://github.com/aspnet/AspNetCore/blob/78b9b2e23e7709cef6659b11e7e0b266f7a1089f/src/Servers/HttpSys/src/RequestProcessing/RequestContext.cs#L55
I would get the correct cancellation token, but I couldn't figure out what condition I should add or remove from if statement, to make this work.

What I'm about to suggest is probably wrong since it changes the result on this test
https://github.com/aspnet/AspNetCore/blob/4a7bf75e89e3dfe7264a641ed041f50788d28d20/src/Servers/HttpSys/test/FunctionalTests/Listener/ServerTests.cs#L49
and may also have other consequences that I'm not aware of, but is it possible to swap `_disposed = true;` and  `FlushInternal(endOfRequest: true);`
https://github.com/aspnet/AspNetCore/blob/320c4b055656b88240cd198b2f40f1f5f0f4ccb3/src/Servers/HttpSys/src/RequestProcessing/ResponseBody.cs#L672-L673

Reply:
If you swap _disposed and FlushInternal you'd have to make sure to handle error conditions from FlushInternal such that _disposed still gets set. E.g. put the dispose in a finally block.

Reply:
Actually that might not be an issue. FlushInternal calls Abort which sets _disposed.